### PR TITLE
feat(ton): integrate new functionality

### PR DESCRIPTION
**Pull request by swift1337**
**Updates**
- [x] Update Gateway code
- [x] Parse and vote for `call` operation
- [x] Support `increase_seqno` outbound method (unblocks tx "cancellation" similarly to SOL and SUI)
- [x] Restrict masterchain `-1` addresses
- [x] Restrict non-compliant outbounds

**E2E** [WIP]
- [x] `call`
- [x] `withdraw_restricted`
- [x] `withdraw_masterchain`

**Related**
- https://github.com/zeta-chain/protocol-contracts-ton/pull/45
- Closes https://github.com/zeta-chain/node/issues/2916
- Closes https://github.com/zeta-chain/protocol-private/issues/226

## Summary by CodeRabbit

- **New Features**
  - Added support for TON contract call operations, enabling cross-chain calls from TON to ZEVM.
  - Introduced compliance checks and restrictions for TON withdrawals to certain addresses and workchains.
  - Implemented new handling for TON sequence number increments with reason codes.
  - Expanded end-to-end tests to cover TON call and restricted withdrawal scenarios.

- **Bug Fixes**
  - Improved error messages and test assertions for better clarity and debugging.

- **Refactor**
  - Streamlined outbound message composition, signature handling, and transaction tracking for TON operations.
  - Enhanced logging and error handling throughout TON-related modules.

- **Tests**
  - Added new test cases and test data for TON call and increase-seqno operations.
  - Improved test coverage for compliance and edge-case scenarios.

- **Chores**
  - Updated configuration and documentation to reflect new TON features and stricter linting rules.

### CODE REVIEW

**Review by lumtis — COMMENTED** [review]
Looking good to me so far outside of the ci failing

File: e2e/e2etests/e2etests.go [hunk 1]
```diff
@@ -855,6 +864,22 @@ var AllE2ETests = []runner.E2ETest{
...
+		TestTONWithdrawRestrictedName,
+		"withdraw TON from ZEVM to restricted address",
+		[]runner.ArgDefinition{
+			{Description: "amount in nano tons", DefaultValue: "100000000"}, // 0.1 TON
+		},
+		TestTONWithdrawRestricted,
+	),
+	runner.NewE2ETest(
+		TestTONWithdrawMasterchainName,
+		"withdraw TON from ZEVM to masterchain",
```
Comment (lumtis) [inline e2e/e2etests/e2etests.go:877]: Can we give more details what does it mean

File: zetaclient/chains/ton/observer/inbound.go [hunk 2]
```diff
@@ -253,17 +257,32 @@
...
+		}
+
+		in.sender = c.Sender
+		in.receiver = c.Recipient
+		in.amount = math.NewUint(0)
+		in.message = c.CallData
+		in.isContractCall = true
+
+		return in, nil
+
```
[the remark this answers is not in the document]
Comment (lumtis) [reply zetaclient/chains/ton/observer/inbound.go:273]: Looks to be legit comment

File: zetaclient/chains/ton/signer/signer_compose.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+	contract "github.com/zeta-chain/node/pkg/contracts/ton"
+	cctypes "github.com/zeta-chain/node/x/crosschain/types"
+	"github.com/zeta-chain/node/zetaclient/compliance"
+)
+
+// CancelReason represents a reason for canceling an outbound (increase_seqno)
+type CancelReason uint32
+
+const (
+	InvalidWorkchain    CancelReason = 1
```
Comment (lumtis) [inline zetaclient/chains/ton/signer/signer_compose.go:16]: Can we add comment what does it mean

File: zetaclient/chains/ton/observer/outbound.go [hunk 4]
```diff
@@ -277,3 +251,25 @@
...
+		if err != nil {
+			return 0, math.Uint{}, errors.Wrap(err, "unable to get increase seqno")
+		}
+
+		// force failure to revert the CCTX in zetacore (similarly to SUI)
+		return chains.ReceiveStatus_failed, cctx.GetCurrentOutboundParam().Amount, nil
+	}
+
+	return 0, math.Uint{}, errors.Errorf("unknown operation %d", o.tx.Operation)
+}
```
[the remark this answers is not in the document]
Comment (swift1337) [reply zetaclient/chains/ton/observer/outbound.go:275]: `_, err := o.tx.IncreaseSeqno()` is actually not needed here

**Review by lumtis — APPROVED** [review]
